### DISCUSSION TITLE
Proxy api requests w/o a custom server

### DISCUSSION THREAD

**Discussion by athrunsun** [vercel/next.js#14057] [Show and tell] [answered]
[Official code example](https://github.com/vercel/next.js/blob/master/examples/with-custom-reverse-proxy) uses a [custom server](https://nextjs.org/docs/advanced-features/custom-server), however by using a custom server, we will lose important performance optimizations, like **serverless functions** and [**Automatic Static Optimization**](https://nextjs.org/docs/advanced-features/automatic-static-optimization). This is NOT acceptable to me.

**Comment by dsauna** [comment]
did you find a solution for this?

**Reply by athrunsun** [reply]
Yes, please take a look at the marked answer. Looks like I misunderstood this type of discussion "show and tell" - I thought it doesn't need to be answered.

Now I edited the question and answered my own question.

**Comment by athrunsun** [comment]
So I went ahead and found a way to proxy api requests to another server with nextjs builtin "[Catch all API routes](https://nextjs.org/docs/api-routes/dynamic-api-routes#catch-all-api-routes)" capability, using a 3rd party library [http-proxy-middleware](https://github.com/chimurai/http-proxy-middleware).

Here is my `src/pages/api/proxy/[...slug].js` file:

```js
import { createProxyMiddleware } from 'http-proxy-middleware';

// Create proxy instance outside of request handler function to avoid unnecessary re-creation
const apiProxy = createProxyMiddleware({
    target: 'http://localhost:5000',
    changeOrigin: true,
    pathRewrite: { [`^/api/proxy`]: '' },
    secure: false,
});

export default function (req, res) {
    apiProxy(req, res, (result) => {
        if (result instanceof Error) {
            throw result;
        }

        throw new Error(`Request '${req.url}' is not proxied! We should never reach here!`);
    });
};
```

So all api requests starting with `/api/proxy` will be forwarded to `http://localhost:5000`.

**Reply by tomsvogel** [reply]
solution works for me, with the right config, had issues with bodyparser on posting forms with files.

`
export const config = {
  api: {
    externalResolver: true,
    bodyParser: false,
  },
};
`
just add this config export.

**Reply by NathanPB** [reply]
@tomsvogel I wish I had paid attention to your comment 2 hours before

**Reply by damonleelcx** [reply]
For some reason it's not working for me and gives me errors like below:
All my server requests start with just /api
```

TypeError: Object(...) is not a function
    at Module../pages/api/[...slug].js (/home/user/app/client/.next/server/pages/api/[...slug].js:109:101)
    at __webpack_require__ (/home/user/app/client/.next/server/pages/api/[...slug].js:23:31)
    at /home/user/app/client/.next/server/pages/api/[...slug].js:91:18
    at Object.<anonymous> (/home/user/app/client/.next/server/pages/api/[...slug].js:94:10)
    at Module._compile (node:internal/modules/cjs/loader:1092:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1121:10)
    at Module.load (node:internal/modules/cjs/loader:972:32)
    at Function.Module._load (node:internal/modules/cjs/loader:813:14)
    at Module.require (node:internal/modules/cjs/loader:996:19)
    at require (node:internal/modules/cjs/helpers:92:18)
    at DevServer.handleApiRequest (/home/user/app/client/node_modules/next/dist/next-server/server/next-server.js:64:181)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:94:5)
    at async Object.fn (/home/user/app/client/node_modules/next/dist/next-server/server/next-server.js:56:492)
    at async Router.execute (/home/user/app/client/node_modules/next/dist/next-server/server/router.js:23:67)
    at async DevServer.run (/home/user/app/client/node_modules/next/dist/next-server/server/next-server.js:66:1042)
```

**Reply by mattvb91** [reply]
Does anyone have a solution for using the provided code by @athrunsun (Which works exactly as needed) but making it usable as an internal API call for `getServerSideProps` without actually doing a fetch? I would like to directly call that method from inside `getServerSideProps` to resolve it from the function?

Currently it only returns a promise that resolves to undefined

**Reply by peppermint-juli** [reply]
THANK YOU SO SO SO MUCH FOR THIS!! I've looking non stop for a decent way to use a proxy with NextJS and this is the only thing that has worked!!!  Complete game changer!

**Reply by jankoritak** [reply]
Does anyone also have an issue getting infinite 301 from the server when going through the proxy? 🤔

**Reply by cancan101** [reply]
Is there anyway to get this to work on a path that doesn't start with /api? For example if I want my proxy at just /proxy?

**Comment by Janpot** [comment, marked as answer]
You can use [custom routes](https://github.com/vercel/next.js/discussions/9081) for this:
```js
rewrites: () => [
  { source: '/api/proxy/:path*', destination: 'http://localhost:5000/:path*' }
]
```

**Reply by athrunsun** [reply]
So it is still experimental? Have you used this technique to proxy requests to external server?

**Reply by Janpot** [reply]
> So it is still experimental?

It's close to being released (https://github.com/vercel/next.js/pull/14602)

> Have you used this technique to proxy requests to external server?

That's one of its intended use cases yes

**Reply by athrunsun** [reply]
Great, I'll give it a try when it's released.

**Reply by dsauna** [reply]
This seems amazing! One problem that I had with the orignial solution is that all the endpoints need to start with /api for the middleware to work. I tried changing all of my calls to /api/whateverMyBackend and then do a rewrite but the problem was that that backend called another endpoint and that one failed. @Janpot with the custom routes I shouldn't have that problem, right?

**Reply by gfazioli** [reply]
I tried but the headers aren't sent, I mean, I would like to "proxy" the cookies as well, any clue?

**Reply by Djerry-dev** [reply]
Using rewrites as a proxy is no longer recommended in Next.js, especially when the project is deployed on Vercel.

The reason is simple: the rewrite mechanism is not a real HTTP proxy. It only rewrites the URL path, and that often causes side effects — weird redirects, missing headers, inconsistent behavior between local and production environments, etc. Next.js was never designed to act as a full proxy layer.

I went through the same issue, and what solved it for me was calling my backend directly with fetch and sending credentials, like this:

```javascript
const data = await fetch("http://localhost:5000", {
  credentials: "include",
});
```

And of course, you need to enable CORS properly on your external backend, including support for credentials.
In my case I was using Fastay.js as the backend, but this works the same with Express, Fastify, Nest, and others.

**Comment by tomdohnal** [comment]
I think that the official example which uses a custom server should be rewritten to the catch-all API routes approach outlined here.
I'm happy to send a PR if this is something you'd accept @timneutkens

**Comment by Birkenstab** [comment]
There is now an official feature for this: Rewrites
Besides normal path rewrites, they can also proxy requests to a different webserver

next.config.js
```javascript
module.exports = {
    async rewrites() {
      return [
        {
          source: '/api/:path*',
          destination: 'http://localhost:5000/:path*' // Proxy to Backend
        }
      ]
    }
  }
```

See [Next.js Docs Rewrites](https://nextjs.org/docs/api-reference/next.config.js/rewrites)

**Reply by edersonsilva** [reply]
> 
> 
> There is now a official feature for this: Rewrites
> Besides normal path rewrites, they can also proxy requests to a different web server
> 
> next.config.js
> 
> ```js
> module.exports = {
>     async rewrites() {
>       return [
>         {
>           source: '/api/:path*',
>           destination: 'http://localhost:5000/:path*' // Proxy to Backend
>         }
>       ]
>     }
>   }
> ```
> 
> See [Next.js Docs Rewrites](https://nextjs.org/docs/api-reference/next.config.js/rewrites)

```js
export async function getStaticPaths() {
  const res = await fetch("/api/products"); <================ It works here ?
  const products = await res.json();

  const paths = products.products.map(product => ({
    params: { id: product.id.toString() }
  }))

  return {
    paths,
    fallback: false
  }
}
```

**Reply by Brazuca228** [reply]
> > There is now a official feature for this: Rewrites
> > Besides normal path rewrites, they can also proxy requests to a different web server
> > next.config.js
> > ```js
> > module.exports = {
> >     async rewrites() {
> >       return [
> >         {
> >           source: '/api/:path*',
> >           destination: 'http://localhost:5000/:path*' // Proxy to Backend
> >         }
> >       ]
> >     }
> >   }
> > ```
> > 
> > 
> > See [Next.js Docs Rewrites](https://nextjs.org/docs/api-reference/next.config.js/rewrites)
> 
> ```js
> export async function getStaticPaths() {
>   const res = await fetch("/api/products"); <================ It works here ?
>   const products = await res.json();
> 
>   const paths = products.products.map(product => ({
>     params: { id: product.id.toString() }
>   }))
> 
>   return {
>     paths,
>     fallback: false
>   }
> }
> ```

> 

> true

**Reply by anoushk1234** [reply]
same here

**Comment by 1ambda** [comment]
Thanks for code snippet above. Was able to write proxy api route w/o HPM. 

Here is my `pages/api/proxy/[...slug].ts` file. Hope this might help others.

```typescript
const proxyConfig: any = {
    logLevel: process.env.NODE_ENV === "production" ? "silent" : "warn",
    target: ENDPOINT_PROXY_API,
    pathRewrite: { "^/api/proxy": "" },
    changeOrigin: true,
    ws: true,
    onError(err, req, res: any) {
        res.writeHead(302, {
            Location: `${ENDPOINT_PROXY_ORIGIN}/exception?message=${err.message}`,
        })
        res.end()
    },
    onProxyReq(proxyReq, req: any, res) {
        /**
         * manually overwrite origin for CORS (changeOrigin might not work)
         */
        proxyReq.setHeader("origin", ENDPOINT_PROXY_ORIGIN)

        const requestId = greq.headers[HEADER_REQUEST_ID] // passed from fetch (Browser) to keep request id in browser too
        const endpoint = req.url
        const method = req.method

        logger.info(`[${PlatformUtil.getLoggerTag()}] (${requestId}) REQ ${endpoint} ${method}`)
    },
    onProxyRes(proxyRes, req: any, res) {
        const requestId = greq.headers[HEADER_REQUEST_ID] // passed from fetch (Browser) to keep request id in browser too
        const endpoint = req.url
        const method = req.method
        const status = proxyRes.statusCode

        logger.info(`[${PlatformUtil.getLoggerTag()}] (${requestId}) RES ${endpoint} ${method} ${status} `)
    },
}

const proxied = createProxyMiddleware(proxyConfig)

export const config = {
    api: {
        // - https://nextjs.org/docs/api-routes/api-middlewares#custom-config
        externalResolver: true,
        bodyParser: false, // not to use url encoded form like streaming POST request
    },
}

export default proxied
```

**Reply by francemj** [reply]
You used HPM when you used `createProxyMiddleware`, didn't you?

**Reply by Linyan-ma** [reply]
i still get "blocked by CORS policy",anyone can help?

**Comment by mpiorowski** [comment]
Rewrites doesnt work when it comes down to webscokets.

**Comment by RacketyWater7** [comment]
> @tomsvogel I wish I had paid attention to your comment 2 hours before
same case. This addition should be added to the above solution so that other people won't have to go through what we did.
